<a href="https://colab.research.google.com/github/Farrukhbala/SSUET-PROJECTS-AI/blob/master/k_Nearest_Neighbors_(k_NN)_Model_with_Scikit_learn_on_Sonar_Data_Project_1_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/jgv7/RandomForest/master/sonar.all-data.csv')


In [178]:
df.head()

0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109  \
0  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
1  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
2  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
3  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
4  0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   

   0.2111 ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084  0.0090  \
0  0.2872 ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049  0.0052   
1  0.6194 ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164  0.0095   
2  0.1264 ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044  0.0040   
3  0.4459 ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048  0.0107   
4  0.3039 ...  0.0045  0.0014  0.0038  0.0013  0.0089  0.0057  0.0027  0.0051   

   0.0032  R  
0  0.0044  R  
1  0.0078  R  
2  0.0117  R  
3  0.0094  R  
4  0.0062  R  

[5 rows x 61 columns]

In [0]:

X = df.drop(columns=['R'])



In [0]:
import numpy as np
y = df['R'].values

for a,i in enumerate(y):
  if i == 'R':
    y[a] = 1
  else:
      y[a] = 0
      
y = np.array(y).astype(int)

#I was getting the error thats why I encoded using pure python

In [0]:
from sklearn.preprocessing import LabelEncoder


In [0]:

# y = LabelEncoder()

In [0]:
# y.fit(Y)
# y_encode = y.transform(Y)

# type(y_encode)

LabelEncoder()

In [0]:
from sklearn.model_selection import train_test_split

#split dataset into train and test data

X_train , X_test , y_train , y_test = train_test_split(X , y , test_size = 0.2 , random_state = 1 , stratify = y)

In [183]:
from sklearn.neighbors import KNeighborsClassifier


#create KNN classifier

knn = KNeighborsClassifier(n_neighbors = 3)

#Fit the classifier to the data

knn.fit(X_train , y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=3, p=2,
           weights='uniform')

In [184]:
# show first 5 model predictions on the test data

knn.predict(X_test)[0:5]

array([0, 1, 1, 1, 0])

In [185]:
#check accuracy of our model on the test data

knn.score(X_test , y_test)

0.8571428571428571

##Using k-Fold Cross-Validation


In [0]:
from sklearn.model_selection import cross_val_score


In [189]:
#Create a new KNN Model
knn_cv = KNeighborsClassifier(n_neighbors=5)

#Train model with cv of 5

cv_scores = cross_val_score(knn_cv , X, y , cv = 15)

#Print each cv score (accuracy) and average them

print(cv_scores)

print('cv_scores mean: {}'.format(np.mean(cv_scores)))

[0.46666667 0.8        0.8        0.86666667 0.33333333 0.93333333
 0.46153846 0.61538462 0.69230769 0.69230769 0.53846154 0.61538462
 0.15384615 0.53846154 0.53846154]
cv_scores mean: 0.6030769230769232


##Our mean score decreased by almost 25 %. So it did not help us.

##Hypertuning model parameters using GridSearchCV
**GridSearchCV works by training our model multiple times on a range of parameters that we specify. That way, we can test our model with each parameter and figure out the optimal values to get the best accuracy results.**

In [0]:
from sklearn.model_selection import GridSearchCV

In [0]:
#create a new knn model

knn2 = KNeighborsClassifier()

In [192]:
#create a dictionary of all values we want to test for n_neighbors

param_grid = {'n_neighbors' : np.arange(1, 25)}


#use grid_search to test all calues for n_neighbors
knn_gscv = GridSearchCV(knn2 , param_grid  , cv = 5)

#fit the model to data

knn_gscv.fit(X, y)

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=5, p=2,
           weights='uniform'),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_neighbors': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24])},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

##Check top peroforming n_neighbors value


In [193]:
knn_gscv.best_params_

{'n_neighbors': 2}

In [194]:
#check mean score for the top performing value of n_neighbors

#We don't have to use that because of decrease in accuracy rate of our model

knn_gscv.best_score_

0.5603864734299517

##We don't have to use that because of decrease in accuracy rate of our model

#Doing normalization on the dataset to improve accuracy.

In [0]:
X = (X-X.mean())/X.std()

In [0]:
from sklearn.model_selection import train_test_split

#split dataset into train and test data

X_train , X_test , y_train , y_test = train_test_split(X , y , test_size = 0.2 , random_state = 1 , stratify = y)

In [200]:
knn = KNeighborsClassifier(n_neighbors = 3)

#Fit the classifier to the data

knn.fit(X_train , y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=3, p=2,
           weights='uniform')

## first 5 model predictions on the test data before **Normalization**

knn.predict(X_test)[0:5]

array([0, 1, 1, 1, 0])


In [201]:
#After normalization of dataset
knn.predict(X_test)[0:5]


array([1, 1, 1, 1, 0])

##Accuracy of our model on the test data before normalization

knn.score(X_test , y_test)

0.8571428571428571

In [0]:
#After normalization accuracy increased by 7 %

knn.score(X_test , y_test)

0.9285714285714286

# Conclusion:
Using many techniques like , k-fold validation , hypertuning the parameters , grid search to increase our model accuracy but all of these techniques did not help but also they decreased the model accuracy by almost 20 to 25 %.

###The final technique was left in my bag was the Normanlization of data. It should have been the first technique to be used on the data but thats how you learn from the mistakes.
##Normalization of data increased the model accurcay by 7 %.